site/content/en/contributions/design/wasm-extension.md

```diff
@@ -0,0 +1,54 @@
+---
+title: "Wasm OCI Image Support"
+---
+
+## Motivation
+Envoy Gateway (EG) should support Wasm OCI image as a remote wasm code source.
+This feature will allow users to deploy Wasm modules from OCI registries, such as Docker Hub, 
+Google Container Registry, and Amazon Elastic Container Registry, to Envoy proxies managed by EG.
+
+## Goals
+* Define the system components needed to support Wasm OCI images as remote Wasm code sources.
+
+## Architecture
+
+![](../wasm-extension.png)
+
+### Control Plane Wasm File Cache
+
+Envoy lacks native OCI image support, therefore, EG needs to download Wasm modules from OIC registries, 
+cache them locally in the file system, and serve them to Envoy over HTTP.
+
+**HTTP Code Source:** For HTTP code source, we have two options: serve Wasm modules directly from their 
+original HTTP URLs, or cache them in EG (as with OCI images). Caching both the HTTP Wasm modules and OCI 
+images inside EG can make UI consistent, for example, sha256sum can be calculated on the EG side and made 
+optional in the API. EG can also periodically pull Wasm modules and update the local file cache.
+
+**Memory Optimization:** Since we cache Wasm modules in the file system, we can optimize the memory usage 
+of the cache and the HTTP server to avoid introducing too much memory consumption by this feature. 
+For example, when receiving a Wasm pulling request form the Envoy, we can open the Wasm file and write 
+the file content directly to response, and then close the file. There won’t be significant caching involved. 
+However, we need to balance the memory usage and the efficiency, which could be addressed in implementation.
+
+**Caching Mechanism:** Cached files will be evicted based on LRU（Last recently used）algorithm. 
```

Comment:
can you add more clarity here ? is there an existing registry cache we can use ?

Reply:
Istio wasm cache fetches wasm module from OCI registry and HTTP url, and saves them to local files. We can reuse most of them and add an HTTP server in front.

https://github.com/istio/istio/tree/master/pkg/wasm